joeynmt/decoders.py

```diff
@@ -424,3 +426,102 @@ def _init_hidden(self, encoder_final: Tensor = None) \
     def __repr__(self):
         return "RecurrentDecoder(rnn=%r, attention=%r)" % (
             self.rnn, self.attention)
+
+
+# pylint: disable=arguments-differ,too-many-arguments
+# pylint: disable=too-many-instance-attributes, unused-argument
+class TransformerDecoder(Decoder):
+    """
+    A transformer decoder with N masked layers.
+    Decoder layers are masked so that an attention head cannot see the future.
+    """
+
+    def __init__(self, num_layers=4, num_heads=8,
+                 hidden_size=512, ff_size=2048, dropout=0.1,
+                 vocab_size=1, freeze=False,
+                 **kwargs):
+        super(TransformerDecoder, self).__init__()
+
+        # build all (num_layers) layers
+        layers = []
+        for _ in range(num_layers):
+            layer = TransformerDecoderLayer(
+                hidden_size,
+                MultiHeadedAttention(num_heads, hidden_size),
+                MultiHeadedAttention(num_heads, hidden_size),
+                PositionwiseFeedForward(hidden_size, ff_size, dropout), dropout)
+            layers.append(layer)
+
+        self.layers = nn.ModuleList(layers)
+        self.norm = nn.LayerNorm(hidden_size)
+        self.pe = PositionalEncoding(hidden_size, dropout=dropout)
+
+        self._hidden_size = hidden_size
+        self._output_size = vocab_size
+
+        self.output_layer = nn.Linear(hidden_size, vocab_size, bias=False)
+
+        if freeze:
+            freeze_params(self)
+
+    def forward(self,
+                trg_embed: Tensor = None,
+                encoder_output: Tensor = None,
+                encoder_hidden: Tensor = None,
+                src_mask: Tensor = None,
+                unrol_steps: int = None,
+                hidden: Tensor = None,
+                trg_mask: Tensor = None,
+                **kwargs):
+        """
+        :param trg_embed: target embeddings
+        :param encoder_output: source representations
+        :param encoder_hidden: unused
+        :param src_mask:
+        :param unrol_steps: unused
+        :param hidden: unused
+        :param trg_mask:
+        :param kwargs:
+        :return:
+        """
+        assert trg_mask is not None, "trg_mask required for Transformer"
+
+        x = trg_embed
+        x = self.pe(x)  # add position encoding
+
```

Comment:
add comment what's happening here: masking padding areas and subsequent positions